Mathlib/Algebra/Group/AddChar.lean

```diff
@@ -31,170 +30,248 @@ see `Mathlib.NumberTheory.LegendreSymbol.AddCharacter`.
+  toFun : A → M
+  /-- The function maps `0` to `1`.
+  Do not use this directly. Instead use `AddChar.map_zero_one`. -/
+  map_zero_one' : toFun 0 = 1
+  /- The function maps addition in `A` to multiplication in `M`.
+  Do not use this directly. Instead use `AddChar.map_add_mul`. -/
+  map_add_mul' : ∀ a b : A, toFun (a + b) = toFun a * toFun b
+
 #align add_char AddChar
 
 end AddCharDef
 
 namespace AddChar
 
--- Porting note(https://github.com/leanprover-community/mathlib4/issues/5020): added
-section DerivedInstances
+section Basic
+-- results which don't require commutativity or inverses
 
-variable (A : Type u) [AddMonoid A] (M : Type v) [CommMonoid M]
+variable {A M : Type*} [AddMonoid A] [Monoid M]
 
-instance : CommMonoid (AddChar A M) :=
-  inferInstanceAs (CommMonoid (Multiplicative A →* M))
+/-- Define coercion to a function. -/
+instance instFunLike : FunLike (AddChar A M) A M where
+  coe := AddChar.toFun
+  coe_injective' φ ψ h := by cases φ; cases ψ; congr
 
-instance : Inhabited (AddChar A M) :=
-  inferInstanceAs (Inhabited (Multiplicative A →* M))
+#noalign add_char.has_coe_to_fun
+
+-- Porting note(https://github.com/leanprover-community/mathlib4/issues/5229): added.
+@[ext] lemma ext (f g : AddChar A M) (h : ∀ x : A, f x = g x) : f = g :=
+  DFunLike.ext f g h
 
-end DerivedInstances
+@[simp] lemma coe_mk (f : A → M)
+    (map_zero_one' : f 0 = 1) (map_add_mul' : ∀ a b : A, f (a + b) = f a * f b) :
+    AddChar.mk f map_zero_one' map_add_mul' = f := by
+  rfl
 
-section CoeToFun
+/-- An additive character maps `0` to `1`. -/
+@[simp] lemma map_zero_one (ψ : AddChar A M) : ψ 0 = 1 := ψ.map_zero_one'
+#align add_char.map_zero_one AddChar.map_zero_one
 
-variable {A : Type u} [AddMonoid A] {M : Type v} [CommMonoid M]
+/-- An additive character maps sums to products. -/
+lemma map_add_mul (ψ : AddChar A M) (x y : A) : ψ (x + y) = ψ x * ψ y := ψ.map_add_mul' x y
+#align add_char.map_add_mul AddChar.map_add_mul
 
 /-- Interpret an additive character as a monoid homomorphism. -/
-def toMonoidHom : AddChar A M → Multiplicative A →* M :=
-  id
+def toMonoidHom (φ : AddChar A M) : Multiplicative A →* M where
+  toFun := φ.toFun
+  map_one' := φ.map_zero_one'
+  map_mul' := φ.map_add_mul'
 #align add_char.to_monoid_hom AddChar.toMonoidHom
 
-open Multiplicative
-
-/-- Define coercion to a function so that it includes the move from `A` to `Multiplicative A`.
-After we have proved the API lemmas below, we don't need to worry about writing `ofAdd a`
-when we want to apply an additive character. -/
-instance instFunLike : FunLike (AddChar A M) A M :=
-  inferInstanceAs (FunLike (Multiplicative A →* M) A M)
-#noalign add_char.has_coe_to_fun
+-- this instance was a bad idea and conflicted with `instFunLike` above
+#noalign add_char.monoid_hom_class
 
-theorem coe_to_fun_apply (ψ : AddChar A M) (a : A) : ψ a = ψ.toMonoidHom (ofAdd a) :=
+@[simp] lemma toMonoidHom_apply (ψ : AddChar A M) (a : Multiplicative A) : ψ.toMonoidHom a = ψ a :=
```

Comment:
Shouldn't the last `a` be of type `A` instead of `Mulitplicative A`? My understanding is that defeq abuse should be avoided in statements.